### PR TITLE
[Model Runner V2] Cache draft logits in model's LM head dtype

### PR DESCRIPTION
**Pull request by TheEpicDolphin**
# Summary
Currently in MRV2 the draft logits are cached in FP32 so that they can be used for rejection sampling. The reason we cache in FP32 is because we are caching the temperature-applied logits, which need the FP32 numerical precision to prevent rounding errors.

We can save 1/2 the memory by simply caching the draft logits in the data type of the model's LM head (typically BF16), and then apply temperature within the rejection sample kernels.

# Test
Added `test_logits_cache_stores_input_logits_bitwise` to check that we are not losing precision during the logits caching in `gumbel_sample`. All tests in `test_rejection_sampler_utils.py` pass.

# Acceptance-rate verification (no regression)
DeepSeek-V4-Flash-DSpark, TP4 GB200, `dspark` / 7 spec tokens / block rejection
sampling / probabilistic draft sampling. SPEED-Bench `throughput_16k`,
`low_entropy`, ISL 2048 / OSL 2048, concurrency 64, 64 warmups + 256 prompts,
`--temperature 1 --top-p 0.95 --ignore-eos`.

I ran 4 arms per commit in two **counterbalanced orders** (set 1 `base,new,base,new`; set 2
`new,base,new,base`) to cancel any drift across run position.

| Run | Commit | Acceptance % | Mean accept len | Output tok/s |
|---|---|---|---|---|
| base_run1 | main | 60.724 | 5.2507 | 6529 |
| base_run2 | main | 58.902 | 5.1232 | 6318 |
| base_run3 | main | 59.535 | 5.1674 | 6577 |
| base_run4 | main | 59.812 | 5.1868 | 6460 |
| new_run1  | #50910 | 60.584 | 5.2409 | 6674 |
| new_run2  | #50910 | 58.668 | 5.1068 | 6506 |
| new_run3  | #50910 | 59.543 | 5.1680 | 6241 |
| new_run4  | #50910 | 60.832 | 5.2582 | 6434 |

**Summary**
| Metric | main | #50910 | Delta |
|---|---|---|---|
| Acceptance rate (mean of 4) | 59.743 % | 59.907 % | **+0.163 pp** |
| Acceptance rate (token-pooled) | 59.738 % | 59.897 % | +0.159 pp |
| Mean acceptance length | 5.1820 | 5.1935 | +0.0114 |
| Accepted / drafted tokens | 2,117,701 / 3,545,010 | 2,118,734 / 3,537,317 | — |
| Same-commit run-to-run spread | 1.822 pp | 2.163 pp | — |
| Welch t / 95% CI | — | — | t = 0.26, CI ±1.53 pp |

No meaningful change in acceptance rates.

### CODE REVIEW

**Review by WoosukKwon — APPROVED** [review]

File: vllm/v1/worker/gpu/spec_decode/speculator.py [hunk 1]
```diff
@@ -131,11 +131,12 @@ def __init__(self, vllm_config: VllmConfig, device: torch.device):
 
         self.draft_logits: torch.Tensor | None = None
         if self.speculative_config.draft_sample_method == "probabilistic":
+            # Pre-temperature logits, cached from the previous decode step.
             self.draft_logits = torch.zeros(
                 self.max_num_reqs,
                 self.num_speculative_steps,
                 self.vocab_size,
-                dtype=torch.float32,
+                dtype=vllm_config.model_config.head_dtype,
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu/spec_decode/speculator.py:139]: nit: Why `head_dtype`?

**Review by WoosukKwon — APPROVED** [review]

File: vllm/v1/worker/gpu/spec_decode/speculator.py (hunk 1 above)
Comment (TheEpicDolphin) [reply vllm/v1/worker/gpu/spec_decode/speculator.py:139]: That matches the dtype of the logits returned by LogitProcessor: https://github.com/vllm-project/vllm/blob/05b7876f50391d5ee5062bf2605d1f33f200272d/vllm/model_executor/layers/logits_processor.py#L115-L135
